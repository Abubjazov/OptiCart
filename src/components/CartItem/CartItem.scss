@import '../../variables';

.cartitem {
	width: 980px;
	height: 250px;
	display: grid;
	column-gap: 10px;
	row-gap: 5px;
	grid-template-columns: 250px 1fr;
	grid-template-areas:
		'header content'
		'header footer';

	header {
		grid-area: header;
		height: 250px;
		width: 250px;
		position: relative;
		overflow: hidden;
		background: $black;
		border: 1px solid $gray-light;
		border-style: none none solid none;

		img {
			width: 100%;
			height: 100%;
		}

		button {
			position: absolute;
			bottom: -50px;
			right: 0;
			left: 0;
			padding: 7px 30px;
			border: none;
			background: $main-color;
			color: $white;
			font-size: 22px;
			line-height: 22px;
			font-weight: 300;
			cursor: pointer;
			transition: all 0.3s linear;
		}

		&:hover {
			img {
				opacity: 0.7;
			}

			button {
				bottom: 100px;
			}
		}
	}

	.cartitem-descriptor {
		padding: 20px;
		grid-area: content;
		height: 195px;
		background: $white;
		border: 1px solid $gray-light;
		border-style: none none solid none;

		h2 {
			color: $main-color;
			font-size: 25px;
			font-weight: 300;
			margin-bottom: 10px;
		}
	}

	footer {
		grid-area: footer;
		display: grid;
		grid-template-columns: 1fr auto auto;
		height: 50px;
		background: $white;
		border: 1px solid $gray-light;
		border-style: none none solid none;

		.quantity {
			font-size: 19px;
			line-height: 19px;
			align-self: center;
			justify-self: start;

			display: grid;
			grid-template-columns: 1fr 1fr 1fr;

			.quantity-value {
				font-size: 19px;
				font-weight: 700;
				align-self: center;
				justify-self: center;
			}

			button {
				background: $white;
				border: none;
				font-size: 30px;
				line-height: 30px;
				font-weight: 300;
				height: 48px;
				width: 60px;
				transition: all 0.3s linear;

				align-self: center;
				justify-self: center;

				&:hover {
					background: $main-color;
					color: $white;
					cursor: pointer;
				}
			}
		}

		.price {
			background: $white;
			color: $black;
			font-size: 18px;
			font-weight: 300;
			line-height: 18px;
			align-self: center;
			justify-self: center;
			padding: 0 30px;
		}

		.full-price {
			background: $main-color;
			color: $white;
			font-size: 22px;
			font-weight: 300;
			line-height: 22px;
			align-self: center;
			justify-self: center;
			padding: 14px 30px;
		}
	}
}
