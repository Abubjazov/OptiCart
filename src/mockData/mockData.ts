import { CartListItem } from '../interfaces'

export const cartItems: CartListItem[] = [
	{
		id: 2,
		product_id: 15,
		name: 'Rangoto Vi',
		picture: 'https://i.ibb.co/T2XSnw0/z9.webp',
		description: 'Nemo dolore officia est natus.',
		price: 11.04,
		quantity: 3,
	},
	{
		id: 3,
		product_id: 16,
		name: 'Armo Sento 2',
		picture: 'https://i.ibb.co/TPT93hh/z10.jpg',
		description: 'Nobis in voluptatum qui voluptate.',
		price: 17.22,
		quantity: 1,
	},
	{
		id: 4,
		product_id: 9,
		name: 'Vertigo 2',
		picture: 'https://i.ibb.co/vvCZ6gp/nike.webp',
		description: 'Architecto omnis sed nostrum aliquid.',
		price: 26.11,
		quantity: 2,
	},
]
