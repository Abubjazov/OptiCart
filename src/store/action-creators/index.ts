/* eslint-disable import/no-anonymous-default-export */
import * as CartActionCreators from './cart.actionCreators'
import * as ProductActionCreators from './product.actionCreators'

export default {
	...CartActionCreators,
	...ProductActionCreators,
}
