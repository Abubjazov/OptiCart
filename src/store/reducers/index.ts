import { combineReducers } from 'redux'
import { cartReducer } from './cartReducer'
import { productReducer } from './productReducer'

export const rootReducer = combineReducers({
	product: productReducer,
	cart: cartReducer,
})

export type RootState = ReturnType<typeof rootReducer>
