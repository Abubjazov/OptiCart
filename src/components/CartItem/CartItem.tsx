import { FC } from 'react'
import { useActions } from '../../hooks/useActions'
import { useTypedSelector } from '../../hooks/useTypedSelector'

import { CartListItem } from '../../interfaces'
import { SmallSpinner } from '../Spinners/SmallSpinner'

import './CartItem.scss'

export const CartItem: FC<CartListItem> = ({
	id,
	name,
	picture,
	description,
	price,
	quantity,
}): JSX.Element => {
	const { currentItemId, status } = useTypedSelector(state => state.cart)
	const { removeFromCart, updateCartQuantity } = useActions()

	const fullPrice = (quantity: number, price: number) => {
		return (quantity * price).toFixed(2)
	}

	const updateQuantity = (e: any) => {
		if (e.target.classList.contains('quantity-minus')) {
			quantity > 1 ? updateCartQuantity(id, quantity - 1) : removeFromCart(id)
		}

		if (e.target.classList.contains('quantity-plus')) {
			updateCartQuantity(id, quantity + 1)
		}
	}

	return (
		<article className='cartitem'>
			<header>
				<img src={picture} alt={name} />
				<button onClick={() => removeFromCart(id)}>
					{status === 'loading' && currentItemId === id ? (
						<SmallSpinner />
					) : (
						'Remove product'
					)}
				</button>
			</header>

			<div className='cartitem-descriptor'>
				<h2>{name}</h2>
				{description}
			</div>

			<footer>
				<div className='quantity'>
					<button className='quantity-minus' onClick={updateQuantity}>
						{status === 'loading' && currentItemId === id ? (
							<SmallSpinner />
						) : (
							'-'
						)}
					</button>
					<span className='quantity-value'>{quantity}</span>
					<button className='quantity-plus' onClick={updateQuantity}>
						{status === 'loading' && currentItemId === id ? (
							<SmallSpinner />
						) : (
							'+'
						)}
					</button>
				</div>

				<h3 className='price'>{price} $</h3>

				<h3 className='full-price'>{fullPrice(quantity, price)} $</h3>
			</footer>
		</article>
	)
}
