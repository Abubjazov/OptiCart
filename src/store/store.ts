import { createStore } from 'redux'
import { rootReducer } from './reducers'

import { composeWithDevTools } from 'redux-devtools-extension'

const preloadedState = {}

export const store = createStore(
	rootReducer,
	preloadedState,
	composeWithDevTools()
)
