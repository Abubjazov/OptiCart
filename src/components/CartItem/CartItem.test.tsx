import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

import { store } from '../../store'
import { mockCartItems } from '../../mockData/mockData'
import { CartItem } from './CartItem'

describe('Component: CartItem', () => {
	test('should render CartItem', () => {
		const { asFragment } = render(
			<Provider store={store}>
				<CartItem {...mockCartItems[0]} />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})
})
