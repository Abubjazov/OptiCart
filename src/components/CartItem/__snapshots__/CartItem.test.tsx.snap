// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CartItem should render CartItem 1`] = `
<DocumentFragment>
  <article
    aria-label="Sneakers Paco in your cart, unit price 54.5 $, quantity 1"
    class="cartitem"
    tabindex="0"
  >
    <header>
      <picture>
        <source
          media="(min-width: 901px)"
          srcset="https://i.ibb.co/DrymCMM/4-250-250.webp"
        />
        <source
          media="(min-width: 651px)"
          srcset="https://i.ibb.co/84RrCYH/4-400-400.webp"
        />
        <source
          media="(max-width: 650px)"
          srcset="https://i.ibb.co/LSd3drp/4-360-360.webp"
        />
        <img
          alt="Paco"
          src="https://i.ibb.co/DrymCMM/4-250-250.webp"
        />
      </picture>
      <button
        aria-label="Remove sneakers Paco from cart"
      >
        Remove product
      </button>
    </header>
    <section
      class="cartitem-descriptor"
    >
      <h2>
        Paco
      </h2>
      <p>
        Ab odit numquam consectetur aut.
      </p>
    </section>
    <footer>
      <div
        class="quantity"
      >
        <button
          aria-label="Reduce the quantity of Paco by one"
          class="quantity-minus"
        >
          -
        </button>
        <h3
          aria-label="Quantity of product Paco is 1"
          class="quantity-value"
          tabindex="0"
        >
          1
        </h3>
        <button
          aria-label="Increase the quantity of Paco by one"
          class="quantity-plus"
        >
          +
        </button>
      </div>
      <h3
        aria-label="Sneakers Paco unit price is 54.5 $"
        class="price"
        tabindex="0"
      >
        54.5 $
      </h3>
      <div
        class="full-price"
      >
        <h3
          aria-label="Total price of the ordered sneakers Paco is 54.50 $"
          tabindex="0"
        >
          54.50 $
        </h3>
      </div>
    </footer>
  </article>
</DocumentFragment>
`;
